squid-deb-proxy 0.8.14 (new formula)

Elemecca opened the pull request:
This adds a new formlua for [squid-deb-proxy], a caching HTTP proxy for Debian / Ubuntu APT repositories based on Squid. It's useful to have on OS X if e.g. you spin up Debian-based VMs frequently with Vagrant.

[squid-deb-proxy]: https://launchpad.net/squid-deb-proxy

Elemecca commented:
Ah, damn, that generated tarball isn't stable.

DomT4 commented:
Use Debian:

```ruby
url "https://mirrors.ocf.berkeley.edu/debian/pool/main/s/squid-deb-proxy/squid-deb-proxy_0.8.14.tar.gz"
```

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,170 @@
+class SquidDebProxy < Formula
+  desc "APT repo caching proxy based on Squid."
+  homepage "https://launchpad.net/squid-deb-proxy"
+
```
Don't need this spacing.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,170 @@
+
+  depends_on "squid"
+
+  # assorted changes that should be submitted upstream
```
Indeed, they should :wink:

DomT4 replied:
Was this patch submitted upstream?

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,170 @@
+      s.gsub! /(\binstall\b.*?)(\s+-[og]\s*\S+)+/, "\\1"
+
+      # fix paths with the HOMEBREW_PREFIX
+      s.gsub! "/usr/sbin/squid", "#{HOMEBREW_PREFIX}/opt/squid/sbin/squid"
```
Ideally `#{sbin}/squid`

Elemecca replied:
Changing this to `#{sbin}/squid` makes the resulting path `/usr/local/Cellar/squid-deb-proxy-0.8.14/sbin/squid`, which is not where `squid` lives. The `squid` binary does not come from this formula; it's inherited from the `squid` formula, thus the dependency.

DomT4 replied:
Oh, aye. Apologies, I misread. It should be:

```ruby
"#{Formula["squid"].opt_sbin}/squid"
```
though.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,170 @@
+        Pathname("mirror-dstdomain.acl.Ubuntu").binread
+    )
+
+    # normal installation called by debian/rules
```
Don't need this comment.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,170 @@
+
+    # pre-installation from debian/rules
+    # we take both sets because we're not either distro
+    Pathname("mirror-dstdomain.acl").write(
```
Don't need to call `Pathname` on these.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,170 @@
+    system "make", "install", "DESTDIR=#{prefix}"
+
+    # additional installation from debian/squid-deb-proxy.install
+    mkdir_p "#{prefix}/usr/share/squid-deb-proxy"
```
This is a bit odd. What is it doing?

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,170 @@
+    cp "init-common.sh", "#{prefix}/usr/share/squid-deb-proxy/"
+
+    # produce execution wrapper script
+    Pathname("#{sbin}/squid-deb-proxy").write(<<-EOS.undent
```
Same comments here and below on `Pathname`

Elemecca commented:
OK, I've addressed most of the review comments.

I'm not sure about the `Pathname` thing, though. I replaced `Pathname.read` and `Pathname.write` with `IO.read` and `IO.write` the first time they were used because it didn't make any real difference. However, in all the other instances of `Pathname` I'm using it because Homebrew has extended `Pathname.write` to automatically create the parent directory, whereas `IO.write` just fails if it doesn't exist. I can switch to explicitly calling `mdir_p( dirname( path ) )` before `IO.write`, but I don't feel that using `Pathname` is wrong there since it actually is doing path manipulation.

Also, out of curiosity, why the switch to the Debian mirror, especially since squid-deb-proxy appears to be an Ubuntu project (at least, it's hosted on Launchpad)?

Elemecca commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,174 @@
+
+    # additional installation from debian/squid-deb-proxy.install
+    mkdir_p "#{prefix}/usr/share/squid-deb-proxy"
+    cp "init-common.sh", "#{prefix}/usr/share/squid-deb-proxy/"
```
> This is a bit odd. What is it doing?

For some reason the upstream project, which is intended to be a Debian package, installs this file using the Debian packaging metadata instead of the `install` target of the Makefile. I therefore have to install it explicitly.

dunn replied:
Could you ask the developers if they'd consider adding it to the Make target?

Also, those two lines can be one:
```ruby
(prefix/"usr/share/squid-deb-proxy").install "init-common.sh"

DomT4 commented:
> Also, out of curiosity, why the switch to the Debian mirror, especially since squid-deb-proxy appears to be an Ubuntu project (at least, it's hosted on Launchpad)?

I presumed this comment:

> Ah, damn, that generated tarball isn't stable.

Was applying to the current URL, which was Launchpad at the time. If that's not the case, use LP by all means.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,174 @@
+    cp "init-common.sh", "#{prefix}/usr/share/squid-deb-proxy/"
+
+    # produce execution wrapper script
+    Pathname.new("#{sbin}/squid-deb-proxy").write(
```
Don't need to call `Pathname` here...

```ruby
(sbin/"squid-deb-proxy").write <<-EOS.undent
  #!/bin/bash

  example
  example 2
  and three
EOS
```
Should work just fine.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,174 @@
+    user_networks  = "#{etc}/squid-deb-proxy/allowed-networks-src.acl.d/10-default"
+    user_dests     = "#{etc}/squid-deb-proxy/mirror-dstdomain.acl.d/10-default"
+    user_blacklist = "#{etc}/squid-deb-proxy/pkg-blacklist.d/10-default"
+    Pathname.new(user_conf).write(
```
Same deal on the `Pathname`. Should work with the same style of syntax example above.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,174 @@
+      # additional Squid configuration
+      EOS
+    )
+    Pathname.new(user_networks).write(
```
Same deal on the `Pathname`. Should work with the same style of syntax example above.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,174 @@
+      #136.199.8.0/24
+      EOS
+    )
+    Pathname.new(user_dests).write(
```
Same deal on the `Pathname`. Should work with the same style of syntax example above.

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,174 @@
+      #repo.steampowered.com
+      EOS
+    )
+    Pathname.new(user_blacklist).write(
```
And here too :wink:

DomT4 commented on Library/Formula/squid-deb-proxy.rb:
```diff
@@ -0,0 +1,174 @@
+    )
+  end
+
+  test do
```
This should be the very last part of the formula (Before the final end and start of the `__END__` patch)

Elemecca commented:
@DomT4 I think I've addressed your comments. Can you re-review? Should I squash down to one commit?

DomT4 commented:
One commit is always nice! Has the patch been submitted upstream yet?

bfontaine commented:
Ping re Dom’s comment?

apjanke commented:
Closing this due to inactivity/lack of response. If you'd still like to pursue this, feel free to respond to the previous comments/questions and re-open on the new [`homebrew/core` repo](https://github.com/homebrew/homebrew-core).
